hcl/eval/functions.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package eval
+
+import (
+	"path/filepath"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/ext/customdecode"
+	tflang "github.com/hashicorp/terraform/lang"
+	"github.com/zclconf/go-cty/cty"
+	"github.com/zclconf/go-cty/cty/function"
+)
+
+func newTmFunctions(basedir string) map[string]function.Function {
+	scope := &tflang.Scope{BaseDir: basedir}
+	tffuncs := scope.Functions()
+
+	tmfuncs := map[string]function.Function{}
+	for name, function := range tffuncs {
+		tmfuncs["tm_"+name] = function
+	}
+
+	// fix terraform broken abspath()
+	tmfuncs["tm_abspath"] = tmAbspath(basedir)
+
+	// sane ternary
+	tmfuncs["tm_ternary"] = tmTernary()
+	return tmfuncs
+}
+
+// tmAbspath returns the `tm_abspath()` hcl function.
+func tmAbspath(basedir string) function.Function {
+	return function.New(&function.Spec{
+		Params: []function.Parameter{
+			{
+				Name: "path",
+				Type: cty.String,
+			},
+		},
+		Type: function.StaticReturnType(cty.String),
+		Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+			path := args[0].AsString()
+			var abspath string
+			if filepath.IsAbs(path) {
+				abspath = path
+			} else {
+				abspath = filepath.Join(basedir, path)
+			}
+
+			return cty.StringVal(filepath.ToSlash(filepath.Clean(abspath))), nil
+		},
+	})
+}
+
+func tmTernary() function.Function {
+	return function.New(&function.Spec{
+		Params: []function.Parameter{
+			{
+				Name: "cond",
+				Type: cty.Bool,
+			},
+			{
+				Name: "val1",
+				Type: customdecode.ExpressionClosureType,
+			},
+			{
+				Name: "val2",
+				Type: customdecode.ExpressionClosureType,
+			},
+		},
+		Type: func(args []cty.Value) (cty.Type, error) {
+			v, err := ternary(args[0], args[1], args[2])
+			if err != nil {
+				return cty.NilType, err
+			}
+			return v.Type(), nil
+		},
+		Impl: func(args []cty.Value, retType cty.Type) (cty.Value, error) {
+			return ternary(args[0], args[1], args[2])
+		},
+	})
+}
+
+func ternary(cond cty.Value, val1, val2 cty.Value) (cty.Value, error) {
+	evalExprVal := func(arg cty.Value) (cty.Value, error) {
+		closure := customdecode.ExpressionClosureFromVal(arg)
+		if dependsOnUnknowns(closure.Expression, closure.EvalContext) {
+			// partial
+			return arg, nil
+		}
+
+		v, diags := closure.Value()
+		if diags.HasErrors() {
+			return cty.NilVal, diags
+		}
+		return v, nil
+	}
```

Comment:
This closure doesn't use anything from its enclosure, why is it just not a function on the package like dependsOnUnknowns ? :thinking: